appcues/src/main/java/com/appcues/ui/ExperienceStepFormState.kt

```diff
@@ -1,39 +1,102 @@
+    private var _formItems: HashMap<UUID, ExperienceStepFormItemState> = hashMapOf()
+
+    private var itemIndex = 0
+
+    val formItems: Collection<ExperienceStepFormItemState>
+        get() = _formItems.values.sortedBy { it.index }
 
     // this can be used by buttons needing to know if they are enabled or not based on required input
     val isFormComplete = mutableStateOf(true)
 
-    fun captureFormItem(id: UUID, item: ExperienceStepFormItem) {
-        formItems[id] = item
-        isFormComplete.value = !formItems.values.any { !it.isComplete }
+    fun getValue(primitive: TextInputPrimitive) =
+        getItem(primitive).value
+
+    fun setValue(primitive: TextInputPrimitive, value: String) {
+        val item = getItem(primitive)
+        item.value = value
+        updateFormItem(item)
+    }
+
+    fun getValue(primitive: OptionSelectPrimitive) =
+        getItem(primitive).values
+
+    fun setValue(primitive: OptionSelectPrimitive, values: Set<String>) {
+        val item = getItem(primitive)
+        item.values = values
+        updateFormItem(item)
+    }
+
+    private fun getItem(primitive: TextInputPrimitive): TextInputFormItemState {
+        var item = _formItems[primitive.id] as? TextInputFormItemState
+        if (item != null) return item
+
+        // create new state tracking object
+        item = with(primitive) {
+            TextInputFormItemState(itemIndex++, id, "textInput", label.text, required, defaultValue ?: "")
+        }
+        updateFormItem(item)
+        return item
+    }
+
+    private fun getItem(primitive: OptionSelectPrimitive): OptionSelectFormItemState {
+        var item = _formItems[primitive.id] as? OptionSelectFormItemState
+        if (item != null) return item
+
+        // create new state tracking object
+        item = with(primitive) {
+            OptionSelectFormItemState(itemIndex++, id, "optionSelect", label.text, required, defaultValue)
+        }
+        updateFormItem(item)
+        return item
+    }
+
+    private fun updateFormItem(item: ExperienceStepFormItemState) {
+        _formItems[item.id] = item
+        isFormComplete.value = !_formItems.values.any { !it.isComplete }
     }
 }
 
-internal sealed class ExperienceStepFormItem(open val label: String, open val isRequired: Boolean) {
+internal sealed class ExperienceStepFormItemState(
+    open val index: Int,
+    open val id: UUID,
+    open val type: String,
+    open val label: String,
+    open val isRequired: Boolean,
+) {
 
     val isComplete: Boolean
         get() {
             return when (this) {
-                is MultipleTextFormItem -> !isRequired || values.isNotEmpty()
-                is SingleTextFormItem -> !isRequired || value.isNotBlank()
+                is OptionSelectFormItemState -> !isRequired || values.isNotEmpty()
+                is TextInputFormItemState -> !isRequired || value.isNotBlank()
             }
         }
 
-    class SingleTextFormItem(
+    class TextInputFormItemState(
```

Comment:
this was renamed, as was the other option, to more explicitly tie to the primitive types that use them.  I felt like it was better to be a bit more clear about that and the general naming wasn't really serving much purpose.